src/backends/mod.rs

```diff
@@ -35,10 +35,205 @@
 // OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 // SOFTWARE.
 
+use errors::HWTracerError;
+use Tracer;
+use backends::dummy::DummyTracer;
+
 #[cfg(perf_pt)]
 pub mod perf_pt;
 #[cfg(perf_pt)]
 pub use self::perf_pt::PerfPTTracer;
 
+#[derive(Debug, PartialEq)]
+pub enum BackendKind {
+    Dummy,
+    PerfPT,
+}
+
+/// Generic configuration interface for all backends.
+/// A `None` field indicates a default value and attributes which don't apply to a given backend
+/// are checked by the `TracerBuilder` when `build()` is called.
+#[derive(Debug)]
+pub struct BackendConfig {
```

Comment:
What's the use in separating this struct from `TracerBuilder`?

Reply:
It get's passed into `from_backend_config()` of the `Tracer` trait, which no longer needs the `backend` attribute.

Reply:
Is it worth the hassle of two structs?

Reply:
Up to you. There's not really any cost in passing the redundant field into `from_backend_config`. The struct is short-lived and that field would be ignored if we did pass it in.

Reply:
I think I'd suggest one struct to keep things simple.

Reply:
Hrm, I'm looking at the struct inlined and thinking "this looks worse than before". We've got the management fields mixed in with the config knobs. Could I convince you to reconsider?

```
 /// A builder interface for instantiating backends.    
  pub struct TracerBuilder {    
      user_backend: Option<BackendKind>,  // User-requested backend choice.    
      auto_backend: BackendKind,          // Auto-chosen backend if `user_backend` is `None`.
      config: BackendConfig,    
      /// Generic backend configuration knobs.    
      /// If a field is `None` at `build()` time then the backend will select a default value. Ant
      /// attributes which don't apply to a given backend are also checked.          
      perf_aux_bufsize: Option<usize>,                                         
      perf_data_bufsize: Option<usize>,                     
      new_trace_bufsize: Option<usize>,    
  }
```

Reply:
It would surely be:

```
  pub struct TracerBuilder {    
      user_backend: Option<BackendKind>,  // User-requested backend choice.         
      perf_aux_bufsize: Option<usize>,                                         
      perf_data_bufsize: Option<usize>,                     
      new_trace_bufsize: Option<usize>,    
  }
```